### PR TITLE
MH-12703: Add userdirectory for Moodle

### PR DESCRIPTION
**Pull request by mtneug**
This PR adds a Moodle user and role provider analogous to the Sakai user and role provider. To efficiently query the necessary information, the [moodle-tool_opencast](https://github.com/unirz-tu-ilmenau/moodle-tool_opencast) plug-in has to be installed in Moodle. ~There is still a minor convenience feature that has to be implemented on Moodle site, which effects the documentation included in this PR. This work is therefore WIP, but I figured that at least the code can be reviewed. Function wise this integration should work.~

### CODE REVIEW

**Comment by mtneug** [comment]
This is now ready for a review.

**Comment by mtneug** [comment]
Any change we might find a reviewer until 22.3 😬? Otherwise I won't be able to work on this until the 5.x feature freeze (but no pressure, if it can't be done for 5.x there is 6.x 🙂).

**Comment by smarquard** [comment]
At a glance, the code looks good to me. It would be ideal to find a reviewer who also uses moodle and can test it in production. But if no-one else volunteers, I'd be prepared to merge it in time for 5.x, as there's no reason it should wait until 6.x and it only really affects people who switch it on, so it's low-risk for everyone else.

File: docs/guides/admin/docs/configuration/security.user.moodle.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+by adding an entry to `OPENCAST/etc/org.ops4j.pax.logging.cfg`:
+
+```
+log4j.logger.org.opencastproject.userdirectory.moodle=DEBUG
+```
+
+### Step 3
+
+You can grant additional roles to all Moodle users in Opencast by creating a
+group with the title 'Moodle'. You can then add additional roles to this group,
```
Comment (wsmirnow) [inline docs/guides/admin/docs/configuration/security.user.moodle.md]: ..with the _name_ 'Moodle'.. (not title)

File: modules/userdirectory-moodle/src/main/java/org/opencastproject/userdirectory/moodle/MoodleWebServiceException.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.opencastproject.userdirectory.moodle;
+
+public class MoodleWebServiceException extends Exception {
```
Comment (wsmirnow) [inline modules/userdirectory-moodle/src/main/java/org/opencastproject/userdirectory/moodle/MoodleWebServiceException.java]: Can you please add documentation for this exception class.

**Review by wsmirnow — CHANGES_REQUESTED** [review]
The code looks good so far and works as expected. Can you please solve the review comments.

**Comment by mtneug** [comment]
I have adapted the PR according to your comments.

**Comment by mtneug** [comment]
@wsmirnow is this good to go or do I need to work on it some more?

**Comment by wsmirnow** [comment]
As there are minor requested changes (documentation of an exception) I suggest to merge this PR and create a patch afterwords. So the feature will be available in 5.x.

**Comment by mtneug** [comment]
@wsmirnow I already added the documentation

**Comment by wsmirnow** [comment]
I've created an other ticket (see MH-12805) for review fixes. I will provide the requested changes there.

**Pull request merged**